### PR TITLE
MAINT improve coverage: openml.datasets.functions

### PR DESCRIPTION
**Pull request by mfeurer**

### CODE REVIEW

**Comment by mfeurer** [comment]
This is a first step to increase the test coverage of the module. Once it's higher I'll start working on the issues with high priority. @amueller do you have time to look at this PR and merge it?

**Comment by amueller** [comment]
please ping me again end of the week. Thinks are crazy here ;)

**Comment by mfeurer** [comment]
Sorry for the delay, things got crazy here, too. Do you have time to review this ATM? I will now continue with dataset.py, task.py and task_functions.py.

File: openml/datasets/functions.py [hunk 1]
```diff
@@ -171,44 +182,44 @@ def check_datasets_active(dids):
...
     active = {}
 
-    dataset_list_idx = 0
     for did in dids:
-        # TODO replace with a more efficient while loop!
-        for idx in range(dataset_list_idx, len(dataset_list)):
-            if did == dataset_list[idx]['did']:
-                active['did'] = bool(dataset_list[idx]['status'])
-        dataset_list_idx = idx
+        for dataset in dataset_list:
```
Comment (amueller) [inline openml/datasets/functions.py]: err this loop seems pretty bad (I know it was there before).
why not iterate over ``dataset_list`` and build a dictionary from all datasets to bool and then subset it with ``dids``?
(it's a bit weird then to actually provide ``dids`` as we always compute it for all).

Comment (amueller) [inline openml/datasets/functions.py]: Maybe that tells us that ``list_datasets`` should actually return a dict indexed by ``did``?

File: openml/datasets/functions.py [hunk 2]
```diff
@@ -236,59 +247,75 @@ def get_dataset(did):
...
-    arff_file = _get_dataset_arff(did, description=description)
+    did_cache_dir = _create_dataset_cache_directory(did)
+
+    try:
+        description = _get_dataset_description(did_cache_dir, did)
+        arff_file = _get_dataset_arff(did_cache_dir, description)
+        # TODO not used yet, figure out what to do with them...
+        features = _get_dataset_features(did_cache_dir, did)
+        qualities = _get_dataset_qualities(did_cache_dir, did)
+    except Exception as e:
```
Comment (amueller) [inline openml/datasets/functions.py]: Be more specific in the exceptions you expect if possible. Not sure though.

File: openml/flows/flow.py [hunk 3]
```diff
@@ -37,9 +37,10 @@ def __init__(self, model, id=None, uploader=None,
         self.tag = tag
         self.model = model
         self.source = "FIXME DEFINE PYTHON FLOW"
+        # TODO add scikit-learn here!
         self.name = (model.__module__ + "." +
                      model.__class__.__name__)
-        self.external_version = 'Tsklearn_' + sklearn.__version__
+        self.external_version = 'sklearn_' + sklearn.__version__
```
Comment (amueller) [inline openml/flows/flow.py]: wasn't the T some openml convention? Not sure though.

File: openml/tasks/functions.py [hunk 4]
```diff
@@ -23,21 +23,19 @@ def _get_cached_tasks():
...
 
     return tasks
 
 
 def _get_cached_task(tid):
     for cache_dir in [config.get_cache_directory(), config.get_private_directory()]:
         task_cache_dir = os.path.join(cache_dir, "tasks")
-        task_file = os.path.join(task_cache_dir,
-                                 "tid_%d.xml" % int(tid))
+        task_file = os.path.join(task_cache_dir, str(tid), "task.xml")
```
Comment (amueller) [inline openml/tasks/functions.py]: so you changed the file names? Or was it broken before? This breaks all the users caches (if anyone uses this) ;)

File: openml/tasks/task.py [hunk 5]
```diff
@@ -29,11 +29,6 @@ def __init__(self, task_id, task_type, data_set_id, target_feature,
         if cost_matrix is not None:
             raise NotImplementedError("Costmatrix")
 
-    def __str__(self):
```
Comment (amueller) [inline openml/tasks/task.py]: Why did you remove this? (it should have been ``__repr__`` I think though)

File: openml/testing.py [hunk 6]
```diff
@@ -14,6 +15,19 @@ class TestBase(unittest.TestCase):
     """
 
     def setUp(self):
+        # This cache directory is checked in to git to simulate a populated
+        # cache
+        self.static_cache_dir = None
+        static_cache_dir = os.path.dirname(os.path.abspath(inspect.getfile(self.__class__)))
```
Comment (amueller) [inline openml/testing.py]: I think you can just use ``__file__`` to refer to the path of this file.

File: openml/testing.py [hunk 7]
```diff
@@ -14,6 +15,19 @@ class TestBase(unittest.TestCase):
...
+        for i in range(2):
```
Comment (amueller) [inline openml/testing.py]: I'm not sure I understand this loop.

File: tests/datasets/test_datasets.py [hunk 8]
```diff
@@ -9,49 +10,65 @@
...
+    def test_get_cached_dataset_description_not_cached(self):
+        openml.config.set_cache_directory(self.static_cache_dir)
+        self.assertRaisesRegexp(OpenMLCacheException, "Dataset description for "
+                                                      "did 3 not cached",
+                                openml.datasets.functions._get_cached_dataset_description,
+                                3)
+
+    def test_get_cached_dataset_arff(self):
+        openml.config.set_cache_directory(self.static_cache_dir)
+        description = openml.datasets.functions._get_cached_dataset_arff(2)
```
Comment (amueller) [inline tests/datasets/test_datasets.py]: maybe ``did=2`` - I like explicit parameter names.

File: tests/tasks/test_task.py [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+        patch.return_value = mock.MagicMock()
+        mm = mock.MagicMock()
+        patch.return_value._retrieve_class_labels = mm
+        patch.return_value._retrieve_class_labels.return_value = 'LA'
+        retval = openml.tasks.get_task(1)
+        self.assertEqual(patch.call_count, 1)
+        self.assertIsInstance(retval, openml.OpenMLTask)
+        self.assertEqual(retval.class_labels, 'LA')
+
+    def test_get_X_and_Y(self):
```
Comment (amueller) [inline tests/tasks/test_task.py]: hmm if we want to follow sklearn api, ``Y`` should be lowercase (because it is usually a vector). Not sure if that was my fault.

**Comment by amueller** [comment]
sorry for the late reply. I'm still not done with my book. I didn't check all the tests in detail but it looks good overall.

**Comment by mfeurer** [comment]
No problem. I will try in incorporate these changes soonish and rebase the current work on that.

File: openml/datasets/functions.py (hunk 2 above)
Comment (mfeurer) [inline openml/datasets/functions.py]: I'm not sure what kind of exceptions can happen. But for all exceptions, the cache dir might be in an undefined state, thus I want to catch them all.

File: openml/datasets/functions.py (hunk 1 above)
Comment (mfeurer) [inline openml/datasets/functions.py]: I'll fix the first comment when I push the next time, and I opened #136 for the second comment.

File: openml/tasks/functions.py (hunk 4 above)
Comment (mfeurer) [inline openml/tasks/functions.py]: The old file name assumed a different directory structure, which is outdated. I think we should only worry about this once we have a serious amount of users?

File: openml/tasks/task.py (hunk 5 above)
Comment (mfeurer) [inline openml/tasks/task.py]: I removed this to had all objects consistent. There is an issue about adding __repr__ to all objects to fix this.

File: openml/testing.py (hunk 6 above)
Comment (mfeurer) [inline openml/testing.py]: No, that makes it crash.

File: openml/testing.py (hunk 7 above)
Comment (mfeurer) [inline openml/testing.py]: Me neither, it'll be gone in the next push.

**Comment by mfeurer** [comment]
Okay, I'll incorporated your suggestions. Can you have a look again?

**Comment by joaquinvanschoren** [comment]
Just Reading through this long thread. Is there something server-side that
I should look at?

Cheers,
Joaquin

**Comment by mfeurer** [comment]
Thanks Joaquin, ubt this only adds new test to the client in order to improve test coverage, so no need for changes on the server side.

**Comment by mfeurer** [comment]
@amueller I pulled in the latest changes from the development branch. Could you please check the code?

File: openml/flows/flow.py (hunk 3 above)
Comment (mfeurer) [inline openml/flows/flow.py]: I checked with Joaquin. No, it wasn't a convention.

File: openml/datasets/functions.py [hunk 10]
```diff
@@ -42,7 +48,8 @@ def _list_cached_datasets():
 
             if "dataset.arff" in dataset_directory_content and \
                     "description.xml" in dataset_directory_content:
-                datasets.append(did)
+                if did not in datasets:
```
Comment (amueller) [inline openml/datasets/functions.py]: how could it already be there? If that was the case, please add a non-regression test checking that after this change, no duplicates appear.

File: openml/datasets/functions.py [hunk 11]
```diff
@@ -176,44 +187,44 @@ def check_datasets_active(dids):
...
-        # TODO replace with a more efficient while loop!
-        for idx in range(dataset_list_idx, len(dataset_list)):
-            if did == dataset_list[idx]['did']:
-                active['did'] = bool(dataset_list[idx]['status'])
-        dataset_list_idx = idx
+        if did not in active:
+            raise ValueError('Could not find dataset %d in OpenML dataset list.'
+                             % did)
+
+    active = {did: active[did] for did in dids}
```
Comment (amueller) [inline openml/datasets/functions.py]: That's ok, though maybe more elegant:
```python
try:
    active = {did: dataset['status'] == 'active'}
except KeyError:
    raise ValueError(ValueError('Could not find dataset %d in OpenML dataset list.' % did)
return active
```

File: openml/datasets/functions.py (hunk 2 above)
Comment (amueller) [inline openml/datasets/functions.py]: alrighty

File: openml/flows/flow.py (hunk 3 above)
Comment (amueller) [inline openml/flows/flow.py]: lol. Why the TODO? that's what the external_version is for, right?

File: openml/runs/run.py [hunk 12]
```diff
@@ -151,6 +155,8 @@ def run_task(task, model):
     train_times = []
 
     rep_no = 0
+    # TODO use different iterator to only provide a single iterator (less
```
Comment (amueller) [inline openml/runs/run.py]: I don't know I understand what that means here.

File: openml/tasks/functions.py (hunk 4 above)
Comment (amueller) [inline openml/tasks/functions.py]: fair

File: openml/tasks/functions.py [hunk 13]
```diff
@@ -65,9 +63,18 @@ def get_estimation_procedure_list():
         "estimationprocedure/list")
     procs_dict = xmltodict.parse(xml_string)
     # Minimalistic check if the XML is useful
-    assert procs_dict['oml:estimationprocedures']['@xmlns:oml'] == \
-        'http://openml.org/openml'
-    assert type(procs_dict['oml:estimationprocedures']['oml:estimationprocedure']) == list
+    if 'oml:estimationprocedures' not in procs_dict:
```
Comment (amueller) [inline openml/tasks/functions.py]: I guess there was some debugging here? This seems pretty heavy handed but it it was useful at some point, maybe it's necessary. I wonder if this can be generalized.

File: openml/testing.py (hunk 6 above)
Comment (amueller) [inline openml/testing.py]: hm... that's pretty odd... but alright.

File: tests/datasets/test_datasets.py [hunk 14]
```diff
@@ -9,49 +10,65 @@
...
-            </oml:authenticate>"""
-        with mock.patch("openml._api_calls._perform_api_call",
-                        return_value=api_mock_return_value) as api_mock:
-            dataset = _get_cached_dataset(2)
-            self.assertIsInstance(dataset, OpenMLDataset)
-            self.assertTrue(api_mock.is_called_once())
+    def test__get_cached_dataset(self, ):
+        openml.config.set_cache_directory(self.static_cache_dir)
+        dataset = _get_cached_dataset(2)
+        self.assertIsInstance(dataset, OpenMLDataset)
```
Comment (amueller) [inline tests/datasets/test_datasets.py]: hm smoketest... not super strict but whatever...

File: tests/tasks/test_task.py [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+        # Classification task
+        task = openml.tasks.get_task(1)
+        X, Y = task.get_X_and_y()
```
Comment (amueller) [inline tests/tasks/test_task.py]: this was renamed... shouldn't that fail now?

Comment (amueller) [inline tests/tasks/test_task.py]: never mind I'm blind.

**Comment by amueller** [comment]
I'm merging this now as it seems a good improvement. My comments still apply ;)

File: openml/datasets/functions.py (hunk 10 above)
Comment (mfeurer) [inline openml/datasets/functions.py]: It could have been there because of line 31, which iterates over two directories. If it was in the 'official' directory, it should not be added again if it is also in the 'private' directory. Because this is not documented and not tested, I opened issue #151 and I will remove this.

File: openml/flows/flow.py (hunk 3 above)
Comment (mfeurer) [inline openml/flows/flow.py]: Removed in 62b68c680a5321f130b4cead43a2aed87522e845

File: openml/runs/run.py (hunk 12 above)
Comment (mfeurer) [inline openml/runs/run.py]: I opened up issue #154

File: openml/tasks/functions.py (hunk 13 above)
Comment (mfeurer) [inline openml/tasks/functions.py]: Hm, a comment would have been useful here. I don't recall why it's there and just leave it until it produces problems.

File: tests/datasets/test_datasets.py (hunk 14 above)
Comment (mfeurer) [inline tests/datasets/test_datasets.py]: what would be the correct way?
